lib/identity_key/identity_key.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <string.h>
+#include <identity_key.h>
+#include <nrf_cc3xx_platform_kmu.h>
+#include <nrf_cc3xx_platform_ctr_drbg.h>
+#include <nrf_cc3xx_platform_identity_key.h>
+#include <hw_unique_key.h>
+#include <zephyr/sys/util.h>
+#include <psa/crypto.h>
+#include <nrfx.h>
+#include <nrfx_nvmc.h>
+
+#include <zephyr/sys/printk.h>
+#include <zephyr/kernel.h>
+
+static int generate_random_secp256r1_private_key(uint8_t *key_buff)
+{
+	psa_status_t status;
+	psa_key_handle_t key_handle;
+	size_t olen;
+
+	/* Initialize PSA Crypto */
+	status = psa_crypto_init();
+	if (status != PSA_SUCCESS) {
+		printk("Failure: Could not initialize PSA crypto!");
+		return -ERR_IDENTITY_KEY_INTERNAL_ERROR;
+	}
+
+	/* Configure the key attributes */
+	psa_key_attributes_t key_attributes = PSA_KEY_ATTRIBUTES_INIT;
+
+	/* Configure the key attributes for Curve type secp256r1
+	 * This key needs to be exported from the volatile storage
+	 */
+	psa_set_key_usage_flags(&key_attributes, PSA_KEY_USAGE_SIGN_HASH | PSA_KEY_USAGE_EXPORT);
+	psa_set_key_lifetime(&key_attributes, PSA_KEY_LIFETIME_VOLATILE);
+	psa_set_key_algorithm(&key_attributes, PSA_ALG_ECDSA(PSA_ALG_SHA_256));
+	psa_set_key_type(&key_attributes, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1));
+	psa_set_key_bits(&key_attributes, IDENTITY_KEY_SIZE_BYTES * 8);
+
+	status = psa_generate_key(&key_attributes, &key_handle);
+	if (status != PSA_SUCCESS) {
+		printk("psa_generate_key failed! Error: %d", status);
+		return -ERR_IDENTITY_KEY_GENERATE_FAILED;
+	}
+
+	status = psa_export_key(key_handle, key_buff, IDENTITY_KEY_SIZE_BYTES, &olen);
+	if (status != PSA_SUCCESS) {
+		printk("psa_export_key failed! Error: %d", status);
+		return -ERR_IDENTITY_KEY_EXPORT_FAILED;
+	}
+
+	if (olen != IDENTITY_KEY_SIZE_BYTES) {
+		printk("Exported key size failed! Key size: %d", olen);
+		return -ERR_IDENTITY_KEY_EXPORT_FAILED;
+	}
+
+	status = psa_destroy_key(key_handle);
+	if (status != PSA_SUCCESS) {
+		printk("psa_destroy_key failed! Error: %d", status);
+		return -ERR_IDENTITY_KEY_INTERNAL_ERROR;
+	}
+
+	return 0;
+}
+
+void identity_key_write_random(void)
+{
+	uint8_t key_buff[IDENTITY_KEY_SIZE_BYTES];
+	int err;
+
+	err = generate_random_secp256r1_private_key(key_buff);
+	if (err != 0) {
+		printk("The identity key generation failed with error code: %d", err);
+		/* Panic raised, as we can't generate the identity key! */
+		k_panic();
+	}
+
+	/* Write the key into KMU using the MKEK as encryption key */
+	identity_key_write_key(key_buff);
+
+	/* Clear the unencrypted key from RAM for security reasons */
+	memset(key_buff, 0x0, IDENTITY_KEY_SIZE_BYTES);
```

Comment:
I'm concerned an optimizing compiler may omit this memset because there is nothing useful that reads what we just wrote here.

Note sometimes memset is built-in so the compiler can make assumptions about what memset does and if it's actually necessary to invoke it. 

Does making key_buff volatile and casting away it's volatility when passing key_buff to the two functions here work?

If not there are a couple of implementations here that might be possible to use:

https://www.cryptologie.net/article/419/zeroing-memory-compiler-optimizations-and-memset_s/

Reply:
Zephyr lacks a standardized function for this, so changing to use a similar approach as Mbed TLS is using (slightly easier to read)

Will reference to a case raised by Mbed TLS with regards to memset_s: 
https://github.com/Mbed-TLS/mbedtls/issues/1584

Fixed